samples/nrf9160/http_update/application_update/src/main.c

```diff
@@ -22,33 +31,343 @@
+}
+
+static int button_init(void)
+{
+	int err;
+
+	if (!device_is_ready(sw0.port)) {
+		printk("SW0 GPIO port device not ready\n");
+		return -ENODEV;
+	}
+
+	err = gpio_pin_configure_dt(&sw0, GPIO_INPUT);
+	if (err < 0) {
+		return err;
+	}
+
+	gpio_init_callback(&sw0_cb, app_download_button_pressed, BIT(sw0.pin));
+
+	err = gpio_add_callback(sw0.port, &sw0_cb);
+	if (err < 0) {
+		printk("Unable to configure SW0 GPIO pin!\n");
+		return err;
+	}
+
+	return 0;
+}
+
+static int apply_state(enum fota_state new_state)
+{
+	__ASSERT(state != new_state, "State already set: %d", state);
+
+	switch (new_state) {
+	case IDLE:
+		app_download_btn_irq_disable();
+		modem_configure_and_connect();
+		break;
+	case CONNECTED:
+		app_download_btn_irq_enable();
+		printk("Press Button 1 or enter 'download' to download new application firmware\n");
+		break;
+	case APP_UPDATE_DOWNLOAD:
+		if (state == APP_UPDATE_APPLY) {
+			printk("Invalid transition: APP_UPDATE_APPLY to APP_UPDATE_DOWNLOAD\n");
+			return -EPERM;
+		}
+		app_download_btn_irq_disable();
+		break;
+	case APP_UPDATE_APPLY:
+		app_download_btn_irq_disable();
+		break;
+	}
+
+	state = new_state;
+
+	return 0;
+}
+
+#if defined(CONFIG_USE_HTTPS)
+static int cert_provision(void)
+{
+	static const char cert[] = {
+		#include "../cert/AmazonRootCA1"
+	};
+	BUILD_ASSERT(sizeof(cert) < KB(4), "Certificate too large");
+
+	int err;
+	bool exists;
+
+	err = modem_key_mgmt_exists(TLS_SEC_TAG,
+				    MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN, &exists);
+	if (err) {
+		printk("Failed to check for certificates err %d\n", err);
+		return err;
+	}
+
+	if (exists) {
+		/* For the sake of simplicity we delete what is provisioned
+		 * with our security tag and reprovision our certificate.
+		 */
+		err = modem_key_mgmt_delete(TLS_SEC_TAG,
+					    MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN);
+		if (err) {
+			printk("Failed to delete existing certificate, err %d\n",
+			       err);
+		}
+	}
+
+	printk("Provisioning certificate\n");
+
+	/*  Provision certificate to the modem */
+	err = modem_key_mgmt_write(TLS_SEC_TAG,
+				   MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN, cert,
+				   sizeof(cert) - 1);
+	if (err) {
+		printk("Failed to provision certificate, err %d\n", err);
+		return err;
+	}
+
+	return 0;
+}
+#endif /* defined(CONFIG_USE_HTTPS) */
+
+/** @brief Configures modem to provide LTE link.
+ *
+ * Blocks until link is successfully established.
+ */
+static int modem_configure_and_connect(void)
+{
+	BUILD_ASSERT(!IS_ENABLED(CONFIG_LTE_AUTO_INIT_AND_CONNECT),
+			"This sample does not support auto init and connect");
+	int err;
+
+#if defined(CONFIG_USE_HTTPS)
+	err = cert_provision();
+	if (err) {
+		printk("Could not provision root CA to %d", TLS_SEC_TAG);
+		return err;
+	}
+
+#endif /* CONFIG_USE_HTTPS */
+
+	printk("LTE Link Connecting ...\n");
+	err = lte_lc_init_and_connect_async(lte_lc_handler);
+	if (err) {
+		printk("LTE link could not be established.");
+		return err;
+	}
+	printk("LTE Link Connected!\n");
```

Comment:
Compiler warning:

../src/main.c: In function 'modem_configure_and_connect':
../src/main.c:280:1: warning: control reaches end of non-void function [-Wreturn-type]
  280 | }

Reply:
Same goes for function in modem_full_update/src/main.c:392

Reply:
Yes, I will push an update shortly.